stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/state/channel/thread/internal/ThreadMutableState.kt

```diff
@@ -18,33 +18,54 @@ package io.getstream.chat.android.offline.plugin.state.channel.thread.internal
 
 import io.getstream.chat.android.client.models.Message
 import io.getstream.chat.android.offline.plugin.state.channel.thread.ThreadState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
 import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
 
-/** State container with mutable data of a thread.*/
-internal interface ThreadMutableState : ThreadState {
-    /** Sorted version of messages. */
-    val sortedMessages: StateFlow<List<Message>>
-
-    /** raw version of messages. */
-    var rawMessages: Map<String, Message>
-
-    /**
-     * Sets the thread to be loading older messages.
-     *
-     * @param isLoading Boolean.
-     */
-    fun setLoadingOlderMessages(isLoading: Boolean)
-
-    /** Sets the end for newer messages. */
-    fun setEndOfOlderMessages(isEnd: Boolean)
-
-    /**
-     * Sets loadings.
-     *
-     * @param isLoading Boolean.
-     */
-    fun setLoading(isLoading: Boolean)
-
-    /** Sets the oldest message in thread. */
-    fun setOldestInThread(message: Message?)
+internal class ThreadMutableState(
+    override val parentId: String,
+    scope: CoroutineScope,
+) : ThreadState {
+    private val _messages = MutableStateFlow(emptyMap<String, Message>())
+    private val _loading = MutableStateFlow(false)
+    private val _loadingOlderMessages = MutableStateFlow(false)
+    private val _endOfOlderMessages = MutableStateFlow(false)
+    private val _oldestInThread: MutableStateFlow<Message?> = MutableStateFlow(null)
+
+    val rawMessage: StateFlow<Map<String, Message>> = _messages
+    override val messages: StateFlow<List<Message>> = _messages
+        .map { it.values }
+        .map { threadMessages -> threadMessages.sortedBy { m -> m.createdAt ?: m.createdLocallyAt } }
+        .stateIn(scope, SharingStarted.Eagerly, emptyList())
+    override val loading: StateFlow<Boolean> = _loading
+    override val loadingOlderMessages: StateFlow<Boolean> = _loadingOlderMessages
+    override val endOfOlderMessages: StateFlow<Boolean> = _endOfOlderMessages
+    override val oldestInThread: StateFlow<Message?> = _oldestInThread
+
+    fun setLoading(isLoading: Boolean) {
+        _loading.value = isLoading
+    }
+
+    fun setLoadingOlderMessages(isLoading: Boolean) {
+        _loadingOlderMessages.value = isLoading
+    }
+
+    fun setEndOfOlderMessages(isEnd: Boolean) {
+        _endOfOlderMessages.value = isEnd
+    }
+
+    fun setOldestInThread(message: Message?) {
+        _oldestInThread.value = message
+    }
+
+    fun deleteMessage(message: Message) {
+        _messages.value -= message.id
```

Comment:
I'm not 100% sure that this will actually delete the message, because this returns a new map, without the key. Although it may remove the key also in the original map (I'm not sure) it think this way would be safer: 
```
_messages.value.toMutableMap().remove(message.id)
```

Reply:
The same goes for `upsertMessages`... using `+=` to make a side effect looks non obvious to me.

Reply:
I understand that this came form the original code and not introduced in this PR, but I think it would be nice to fix it here, since it is a very small change.

Reply:
Nevermind, we are reatributing the value, this is necessary.